File: truevfs-ext/truevfs-ext-insight/src/main/java/net/java/truevfs/ext/insight/I5tOutputStream.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.insight;

import net.java.truevfs.comp.inst.InstrumentingOutputStream;
import net.java.truevfs.comp.jmx.JmxComponent;

import javax.annotation.WillCloseWhenClosed;
import javax.annotation.concurrent.NotThreadSafe;
import java.io.IOException;
import java.io.OutputStream;

/**
 * A controller for an {@link java.io.OutputStream}.
 *
 * @author Christian Schlichtherle
 */
@NotThreadSafe
final class I5tOutputStream extends InstrumentingOutputStream<I5tMediator> implements JmxComponent {

    I5tOutputStream(I5tMediator mediator, @WillCloseWhenClosed OutputStream out) {
        super(mediator, out);
    }

    @Override
    public void activate() {
    }

    @Override
    public void write(final int b) throws IOException {
        final var start = System.nanoTime();
        out.write(b);
        mediator.logWrite(System.nanoTime() - start, 1);
    }

    @Override
    public void write(final byte[] b, final int off, final int len) throws IOException {
        final var start = System.nanoTime();
        out.write(b, off, len);
        mediator.logWrite(System.nanoTime() - start, len);
    }
}

File: truevfs-ext/truevfs-ext-insight/src/main/scala/net/java/truevfs/ext/insight/I5tOutputStream.scala
/*
 * Copyright (C) 2005-2015 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.ext.insight

import java.io._
import javax.annotation._
import javax.annotation.concurrent._
import net.java.truevfs.comp.inst._
import net.java.truevfs.comp.jmx._

/**
 * A controller for an [[java.io.OutputStream]].
 *
 * @author Christian Schlichtherle
 */
@NotThreadSafe
private final class I5tOutputStream(
  mediator: I5tMediator, @WillCloseWhenClosed out: OutputStream
) extends InstrumentingOutputStream(mediator, out) with JmxComponent {

  override def activate(): Unit = { }

  override def write(b: Int): Unit = {
    val start = System.nanoTime
    out write b
    mediator logWrite (System.nanoTime - start, 1)
  }

  override def write(b: Array[Byte], off: Int, len: Int): Unit = {
    val start = System.nanoTime
    out write (b, off, len)
    mediator logWrite (System.nanoTime - start, len)
  }
}
